centaur/src/main/scala/centaur/json/JsonUtils.scala

```diff
@@ -26,25 +26,43 @@ object JsonUtils {
       * The second handles scatters, where each scatter has its own element in the array. In this case all of the
       * scatters are flattened themselves with the scatter index being inserted into the flattened key name.
       */
-    def flatten(prefix: String = ""): JsObject = jsValue.asJsObject.fields.foldLeft(JsObject.empty) {
-      case (acc, (k, v: JsArray)) if v.isSingleCallArray => acc ++ JsObject(k -> v.elements.head).flatten(prefix)
-      case (acc, (k, v: JsArray)) if v.hasShardIndex =>
-        // The .get on the shardIndex is safe as we know all elements of the array have a shardIndex field
-        acc ++ v.elements.map(_.asJsObject).fold(JsObject.empty) { (x, y) => x ++ y.flatten(s"$k.${y.shardIndex.get}") }
-      case (acc, (k, v: JsArray)) =>
-        v.elements.zipWithIndex.foldLeft(acc) { case (accumulator, (element, idx)) =>
-          val maybePrefix = if (prefix.isEmpty) "" else s"$prefix."
-          element match {
-            case _: JsObject => accumulator.mergeWith(element.flatten(s"$maybePrefix$k.$idx"))
-            case x: JsValue => accumulator + (s"$maybePrefix$k.$idx" -> x)
+    def flatten(prefix: String = ""): JsObject = {
+
+      def flattenShardAndAttempt(k: String, v: JsArray, f: JsObject => String): JsObject = {
+        v.elements.map(_.asJsObject).fold(JsObject.empty) { (x, y) => x ++ y.flatten(s"$k.${f(y)}") }
+      }
+
+      jsValue.asJsObject.fields.foldLeft(JsObject.empty) {
+        case (acc, (k, v: JsArray)) if v.isSingleCallArray => acc ++ JsObject(k -> v.elements.head).flatten(prefix)
+        case (acc, (k, v: JsArray)) if v.hasShardIndex && v.hasAttemptNumber =>
+          /* The .get on the shardIndex and attemptNumber is safe as we know all elements of the array have a
+             shardIndex and attempt field. This conversion will add the attempt number in the flattened key structure
+             to avoid lossy conversion for multiple attempts of the same shard
+          */
+          acc ++
+            flattenShardAndAttempt(k, v, (y: JsObject) => y.shardIndex.get) ++
```

Comment:
If you moved the `.` from line 32's `y.flatten(s"$k.${f(y)}")` here instead, you could have an extra line at the top here:
```scala
flattenShardAndAttempt(k, v, (y: JsObject) => "") ++
flattenShardAndAttempt(k, v, (y: JsObject) => s".${y.shardIndex.get}") ++
flattenShardAndAttempt(k, v, (y: JsObject) => s".${y.shardIndex.get}.${y.attemptNumber.get}")
```

Which would mean we could remove the special case handler at the top, and any test could choose whichever of the three it preferred:
```
case (acc, (k, v: JsArray)) if v.isSingleCallArray => acc ++ JsObject(k -> v.elements.head).flatten(prefix)
```

As well as feeling a bit neater, that would have the benefit of stopping tests from failing automatically if they needed more than one attempt (it would at least be random which attempt information they got rather than guaranteed to be missing - and if only one attempt had the field in question, it would always succeed)

Reply:
If we want to move the `.` from line 32's `y.flatten(s"$k.${f(y)}")` to below, won't the modified code be as below? This is because in the special case where the array only has one element we recursively call the flatten with prefix and not empty string?

```
def flattenShardAndAttempt(v: JsArray, f: JsObject => String): JsObject = {
        v.elements.map(_.asJsObject).fold(JsObject.empty) { (x, y) => x ++ y.flatten(f(y)) }
 }
 ....
flattenShardAndAttempt(v, (y: JsObject) => prefix) ++
flattenShardAndAttempt(v, (y: JsObject) => s"$k.${y.shardIndex.get}") ++
flattenShardAndAttempt(v, (y: JsObject) => s"$k.${y.shardIndex.get}.${y.attemptNumber.get}")
```

Reply:
So there is an issue with the code I suggested above. It seems when there is a single array, we loose the key details if we use `flattenShardAndAttempt(v, (y: JsObject) => prefix)`. Using this produces keys like `calls.executionStatus` instead of `calls.<wf_name>.<task_name>.executionStatus`. We need the original special case to add the key details in the flattened structure. So I will be adding that case back. This complicated logic can be revisited in the [BW-483](https://broadworkbench.atlassian.net/browse/BW-483).